File: solutions/0/0/3/6/p0036_s0.c
/* Solution to Euler Problem 36: Double-base Palindromes. */
#include "runner.h"

/* Test binary palindromicity by extracting bits (LSB first) and mirror-comparing; O(log n). */
static int is_binary_palindrome(long long n) {
    if (n <= 0) return 0;
    char bits[64];
    int len = 0;
    long long tmp = n;
    while (tmp > 0) {
        bits[len++] = (char)(tmp & 1);
        tmp >>= 1;
    }
    /* bits is stored LSB first, so check if it reads the same forwards and backwards */
    for (int i = 0; i < len / 2; i++) {
        if (bits[i] != bits[len - 1 - i]) return 0;
    }
    return 1;
}

/* Integer 10^exp, kept exact to avoid floating-point rounding in the half-loop bound. */
static long long int_pow10(int exp) {
    long long result = 1;
    for (int i = 0; i < exp; i++) result *= 10;
    return result;
}

/* Generate decimal palindromes from their left half, then keep the binary palindromes;
   O(sqrt(N) log N). */
const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int max_digits = parse_int(argv[1]);
    long long total = 0;

    /* Single-digit palindromes: 1-9 */
    for (int digit = 1; digit <= 9; digit++) {
        if (is_binary_palindrome((long long)digit)) {
            total += digit;
        }
    }

    /* Multi-digit palindromes generated from left half */
    long long limit = int_pow10(max_digits / 2);
    for (long long digits = 1; digits < limit; digits++) {
        /* Convert digits to string */
        char digits_str[20];
        snprintf(digits_str, sizeof(digits_str), "%lld", digits);
        int num_digits = (int)strlen(digits_str);

        /* Build reversed string */
        char digits_rev[20];
        for (int i = 0; i < num_digits; i++) {
            digits_rev[i] = digits_str[num_digits - 1 - i];
        }
        digits_rev[num_digits] = '\0';

        /* Even-length palindrome: digits_str + digits_rev */
        char even_str[40];
        snprintf(even_str, sizeof(even_str), "%s%s", digits_str, digits_rev);
        long long even_num = atoll(even_str);
        if (is_binary_palindrome(even_num)) {
            total += even_num;
        }

        /* Odd-length palindromes: digits_str + mid_digit + digits_rev (interior 0 is legal) */
        if (2 * num_digits < max_digits) {
            for (int mid = 0; mid <= 9; mid++) {
                char odd_str[42];
                snprintf(odd_str, sizeof(odd_str), "%s%d%s", digits_str, mid, digits_rev);
                long long odd_num = atoll(odd_str);
                if (is_binary_palindrome(odd_num)) {
                    total += odd_num;
                }
            }
        }
    }

    { snprintf(_answer, sizeof _answer, "%lld", (long long)(total)); return _answer; }
}

File: solutions/0/0/3/6/p0036_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 36: Double-base Palindromes [Level 0]. """
from __future__ import annotations

import typing

from solver.runners import runner


def generate_decimal_palindromes(max_digits: int) -> typing.Generator[int, None, None]:
    """Yield every decimal palindrome up to max_digits by mirroring each left half exactly once."""
    for digit in range(1, 10):
        yield digit
    for digits in range(1, 10 ** (max_digits // 2)):
        digits_str = str(digits)
        digits_rev = digits_str[::-1]
        num_digits = len(digits_str)
        yield int(digits_str + digits_rev)
        if 2 * num_digits < max_digits:
            for mid_digit in "0123456789":
                yield int(digits_str + mid_digit + digits_rev)


@runner.main
def solve(*args: str) -> str:
    """Generate decimal palindromes from their left half, then keep the binary palindromes; O(sqrt(N) log N)."""
    max_digits = runner.parse_int(args[0])

    return str(sum(
        (
            number
            for number in generate_decimal_palindromes(max_digits)
            if number == int(str(bin(number))[2:][::-1], base=2)
        )
    ))


if __name__ == "__main__":
    raise SystemExit(solve())
